### PR TITLE
feat: emit multipe parameters

### PR DESCRIPTION
**Pull request by SSebo**
js client can emit multiple parameters in one call. user socketio server `socket.on("foo", async (userId, body, callback)=>{})`  (#246) accept this kind of data which rust-socektio can not emit.

### CODE REVIEW

**Comment by david-sailplan** [comment]
Seems to be working as expected for the most part but I seem to be running into issues when the Value is a string. Namely it seems to not take that value as a singular argument, but instead as an array of characters and sends each character as a different argument.

Of course, I might be doing something wrong here, so please correct me if I am

For example lets say I have 
```
socket.on("test", async(arg1, arg2, arg3, callback)=>{
   console.log([arg1, arg2, arg3])
});
```

The following will work as expected 
```
let args = vec![json!({"foo": "foo"}), json!("{bar": "bar"}), json!({"foobar": "foobar"})]
client.emit_multi_with_ack("test", args, duration, callback);
```
Result
```
[{"foo": "foo"}, {"bar": "bar"}, {"foobar": "foobar"}]
```


But if you try to send a string it seems to split the arguments into the characters
```
let args = vec![json!("foo"), json!("bar"), json!("foobar")]
client.emit_multi_with_ack("test", args, duration, callback);
```

Results in the following
```
['f', 'o', 'o']
```

File: socketio/examples/callback.rs [hunk 1]
```diff
@@ -13,9 +13,10 @@ fn main() {
         match payload {
             Payload::String(str) => println!("Received string: {}", str),
             Payload::Binary(bin_data) => println!("Received bytes: {:#?}", bin_data),
+            Payload::Number(num) => println!("Received number: {}", num),
```
Comment (nshaaban-cPacket) [inline socketio/examples/callback.rs:16]: What value does this type add? Isn't it already covered by binary?

File: socketio/src/client/raw_client.rs [hunk 2]
```diff
@@ -200,9 +236,86 @@ impl RawClient {
...
+    ///    }
+    /// };
+    ///
+    /// let payload = json!({"token": 123});
+    /// socket.emit_with_ack("foo", payload, Duration::from_secs(2), ack_callback).unwrap();
+    ///
+    /// sleep(Duration::from_secs(2));
+    /// ```
+    #[inline]
+    pub fn emit_multi_with_ack<F, E, D>(
```
Comment (nshaaban-cPacket) [inline socketio/src/client/raw_client.rs:298]: What value does this add? Could we add and `impl From Vec<...> for Payload `, so the existing function works?

File: socketio/src/socket.rs [hunk 3]
```diff
@@ -70,54 +70,77 @@ impl Socket {
...
-
-                Ok(Packet::new(
-                    PacketId::Event,
-                    nsp.to_owned(),
-                    Some(payload),
-                    id,
-                    0,
-                    None,
-                ))
+        let mut attachments = vec![];
```
Comment (nshaaban-cPacket) [inline socketio/src/socket.rs:95]: nit/suggestion: code is hard to follow

File: socketio/src/packet.rs [hunk 4]
```diff
@@ -118,24 +118,8 @@ impl From<&Packet> for Bytes {
...
-                format!(
-                    "[{{\"_placeholder\":true,\"num\":{}}}]",
-                    packet.attachment_count - 1,
-                )
-            };
 
-            // build the buffers
-            buffer.put(placeholder.as_ref());
-        } else if let Some(data) = packet.data.as_ref() {
+        if let Some(data) = packet.data.as_ref() {
```
Comment (nshaaban-cPacket) [inline socketio/src/packet.rs:122]: these two are not equivalent, what changed?

File: socketio/examples/readme.rs [hunk 5]
```diff
@@ -10,6 +10,7 @@ fn main() {
         match payload {
             Payload::String(str) => println!("Received: {}", str),
             Payload::Binary(bin_data) => println!("Received bytes: {:#?}", bin_data),
+            Payload::Number(num) => println!("Received: {}", num),
```
Comment (nshaaban-cPacket) [inline socketio/examples/readme.rs:13]: nit: what value does this add?

**Review by nshaaban-cPacket — COMMENTED** [review]
Thanks for the PR! Don't quite understand what the problem was, nor the solution, a brief explanation (preferably in comments) as to what happened may be helpful

**Comment by david-sailplan** [comment]
I had left an earlier comment but deleted it. Turns out it was an issue on the server side

**Comment by david-sailplan** [comment]
This does seem to solve the problem I was experiencing [here](https://github.com/1c3t3a/rust-socketio/issues/246#issuecomment-1286424751). Good job

File: socketio/examples/callback.rs (hunk 1 above)
Comment (SSebo) [reply socketio/examples/callback.rs:16]: currently rust_socketio can not emit `number` value, it will encode as string

File: socketio/src/client/raw_client.rs (hunk 2 above)
Comment (SSebo) [reply socketio/src/client/raw_client.rs:298]: This function wants to emits `Vec<Payload>` as what javascript version of socketio client did.

File: socketio/src/packet.rs (hunk 4 above)
Comment (SSebo) [reply socketio/src/packet.rs:122]: currently only support emit one binary payload, if we want to emit multiple binary payload and keep the same order user emit, placeholders are hard to replace in `packet.rs`. so this logic is moved to `socket.rs`

File: socketio/examples/readme.rs (hunk 5 above)
Comment (SSebo) [reply socketio/examples/readme.rs:13]: currently can not emit simple number payload

**Comment by 1c3t3a** [comment]
First of all thanks for the quick solution @SSebo! But I wanted to ask whether this can be solved by reverting the behaviour of #220. This was the initial reason why I only accepted valid json and nothing else. I would prefer this approach over a completely new API call!

**Comment by 1c3t3a** [comment]
This is not entirely true, we don't need number since this is part of Json, eg: `{ "my_number": 1}`. Can you confirm that a revert of #220 results in being able to emit to a socketio server with `socket.on("foo", async (userId, body, callback)=>{})`where body is not empty? :)

**Comment by SSebo** [comment]
> 

I agree, not include number, I removed Payload::Number part.

**Comment by SSebo** [comment]
> This is not entirely true, we don't need number since this is part of Json, eg: `{ "my_number": 1}`. Can you confirm that a revert of #220 results in being able to emit to a socketio server with `socket.on("foo", async (userId, body, callback)=>{})`where body is not empty? :)

@david-sailplan can you confirm that use a single json payload is able to emit to a socketio server with `socket.on("foo", async (userId, body, callback)=>{})`where body is not empty? :)

**Comment by david-sailplan** [comment]
@SSebo I'm not entirely sure what you're asking here. If you're asking if something to the effect of 
```
let payload = json!(("user_id_1", json!({'data": 1})))
socket.emit_with_ack("foo",payload, duration, callback) 
```
works with server code
```
socket.on("foo", async (userId, body, callback)=>{
    console.log({userId, body, callback});
    callback(1);
})
```

I can confirm that it does not. Log would be 
```
{
    userId: ['user_id)1', {'data': 1}]
    body: [Function (anonymous)],
    callback: undefined
}
```

**Comment by david-sailplan** [comment]
If what you're asking is if 
```
let payload = vec![json!(1), json({"data": 2}))]
socket.emit_multi_with_ack("foo", payload, duration, callback)
```
works with the above server code the answer is yes. It behaves as we expected even after removing the Payload::Number

File: socketio/src/socket.rs [hunk 6]
```diff
@@ -70,52 +70,85 @@ impl Socket {
...
+        data += "]";
+
+        (data, attachments)
+    }
+
+    fn encode_payloads(data: &mut String, payloads: Vec<Payload>, attachments: &mut Vec<Bytes>) {
+        for (index, payload) in payloads.iter().enumerate() {
+            match payload {
+                Payload::Binary(bin_data) => {
+                    *data += "{\"_placeholder\":true,\"num\":";
```
Comment (nshaaban-cPacket) [inline socketio/src/socket.rs:136]: nit: should this number of attachments be done at the top level?

File: socketio/src/socket.rs [hunk 7]
```diff
@@ -70,52 +70,85 @@ impl Socket {
...
+        let mut data = "[".to_owned();
+
+        if let Some(event) = event {
+            data += &format!("\"{}\"", String::from(event));
+            if !payloads.is_empty() {
+                data += ","
+            }
+        }
+
+        Self::encode_payloads(&mut data, payloads, &mut attachments);
```
Comment (nshaaban-cPacket) [inline socketio/src/socket.rs:125]: nit: seems to be re-implementing a json serialization. Would storing in a temporary struct then use serde_json's to_string be simpler?

File: socketio/src/client/client.rs [hunk 8]
```diff
@@ -122,6 +157,58 @@ impl Client {
...
+    ///         Payload::Binary(bytes) => println!("Received bytes: {:#?}", bytes),
+    ///    }
+    /// };
+    ///
+    /// let payload = vec![json!({"token": 123})];
+    /// socket.emit_multi_with_ack("foo", payload, Duration::from_secs(2), ack_callback).unwrap();
+    ///
+    /// sleep(Duration::from_secs(2));
+    /// ```
+    pub fn emit_multi_with_ack<F, E, D>(
```
Comment (nshaaban-cPacket) [inline socketio/src/client/client.rs:195]: nit: we may be able to avoid the need for a new function by modifying the payload type, perhaps adding a Multi variant to payload?

**Review by nshaaban-cPacket — COMMENTED** [review]
Just dropping by, as always @1c3t3a 's in charge

File: socketio/src/payload.rs [hunk 9]
```diff
@@ -1,31 +1,45 @@
...
-/// used for both representing data that's send and data that's received.
+/// The enum is used for both representing data that's send and
+/// data that's received.
 #[derive(Debug, PartialEq, Eq, Clone)]
+#[non_exhaustive]
 pub enum Payload {
     Binary(Bytes),
-    String(String),
+    Json(Value),
+    Multi(Vec<RawPayload>),
```
Comment (nshaaban-cPacket) [inline socketio/src/payload.rs:12]: Not sure if this is better, or if this type even should be recursive, another option would be:
```suggestion
    Multi(Vec<Box<Payload>>),
```

File: socketio/src/socket.rs [hunk 10]
```diff
@@ -70,54 +87,75 @@ impl Socket {
...
     /// Returns a packet for a payload, could be used for bot binary and non binary
     /// events and acks. Convenance method.
     #[inline]
-    pub(crate) fn build_packet_for_payload<'a>(
-        &'a self,
+    pub(crate) fn build_packet_for_payload(
         payload: Payload,
-        event: Event,
-        nsp: &'a str,
+        event: Option<Event>,
```
Comment (nshaaban-cPacket) [inline socketio/src/socket.rs:101]: nit: when is event none? Only usage I can see is in the tests

File: socketio/src/socket.rs [hunk 11]
```diff
@@ -16,6 +18,21 @@ pub(crate) struct Socket {
     connected: Arc<AtomicBool>,
 }
 
+#[derive(Serialize)]
+struct BinaryPlaceHolder {
```
Comment (nshaaban-cPacket) [inline socketio/src/socket.rs:22]: :heart:

File: socketio/src/packet.rs [hunk 12]
```diff
@@ -234,27 +220,12 @@ impl TryFrom<&Bytes> for Packet {
...
-                if str.is_empty() {
-                    packet.data = None
-                } else {
-                    packet.data = Some(str)
-                }
-            }
-            _ => packet.data = Some(json_data.to_string()),
+        packet.data = match json_data {
+            Value::Array(vec) if vec.is_empty() => None,
+            // Value::Array(vec) if vec.len() == 1 => vec.get(0).map(|v| v.to_owned()),
```
Comment (nshaaban-cPacket) [inline socketio/src/packet.rs:227]:
```suggestion
```

File: socketio/src/packet.rs [hunk 13]
```diff
@@ -234,27 +220,12 @@ impl TryFrom<&Bytes> for Packet {
...
+            _ => Some(json_data),
```
Comment (nshaaban-cPacket) [inline socketio/src/packet.rs:228]: nit: would this still have the `_placeholder` in it?

File: socketio/src/packet.rs [hunk 14]
```diff
@@ -21,7 +21,7 @@ pub enum PacketId {
 pub struct Packet {
     pub packet_type: PacketId,
     pub nsp: String,
-    pub data: Option<String>,
+    pub data: Option<Value>,
```
Comment (nshaaban-cPacket) [inline socketio/src/packet.rs:24]: blocker: This is a public type, changing this would be a breaking change. While I agree data: Option<Value> makes the internal implementation more elegant, not worth introducing more breaking changes. Changing the payload type on a PATCH IMO is already pushing the boundary of what should be done on a PATCH even in 0.X.X
```suggestion
    // FIXME: on next breaking change consider changing this to a Value
    pub data: Option<String>,
```

File: socketio/src/payload.rs [hunk 15]
```diff
@@ -1,31 +1,45 @@
...
-/// `Payload::String` which holds a [`std::string::String`]. The enum is
-/// used for both representing data that's send and data that's received.
+/// The enum is used for both representing data that's send and
+/// data that's received.
 #[derive(Debug, PartialEq, Eq, Clone)]
+#[non_exhaustive]
 pub enum Payload {
     Binary(Bytes),
-    String(String),
+    Json(Value),
```
Comment (nshaaban-cPacket) [inline socketio/src/payload.rs:11]: nit: consider adding a way to have some sort of way to convert a struct that impls Deserialize into Payload.

File: socketio/src/client/raw_client.rs [hunk 16]
```diff
@@ -306,51 +319,95 @@ impl RawClient {
...
+            },
+            _ => Event::Message,
         };
 
-        if let Some(attachments) = &packet.attachments {
-            if let Some(binary_payload) = attachments.get(0) {
-                self.callback(&event, Payload::Binary(binary_payload.to_owned()))?;
+        let payload = match &packet.data {
+            Some(Value::Array(vec)) if vec.len() == 1 => {
+                let value = vec.get(0).unwrap();
```
Comment (nshaaban-cPacket) [inline socketio/src/client/raw_client.rs:333]:
```suggestion
                // SAFETY: length checked above
                let value = vec.get(0).unwrap();
```

File: socketio/src/client/raw_client.rs [hunk 17]
```diff
@@ -306,51 +319,95 @@ impl RawClient {
...
+                _ => Event::Message,
+            },
+            _ => Event::Message,
         };
 
-        if let Some(attachments) = &packet.attachments {
-            if let Some(binary_payload) = attachments.get(0) {
-                self.callback(&event, Payload::Binary(binary_payload.to_owned()))?;
+        let payload = match &packet.data {
+            Some(Value::Array(vec)) if vec.len() == 1 => {
```
Comment (nshaaban-cPacket) [inline socketio/src/client/raw_client.rs:332]: nit: clippy lint https://rust-lang.github.io/rust-clippy/master/#len_zero
```suggestion
            Some(Value::Array(vec)) if ! vec.is_empty() => {
```

File: socketio/src/client/raw_client.rs [hunk 18]
```diff
@@ -306,51 +319,95 @@ impl RawClient {
...
+                        serde_json::Value::String(ev) => ev,
+                        _ => "message",
+                    })
+                    .unwrap_or("message")
+                    .into()
+            } else {
+                Event::Message
+            };
+
+            let payload = if contents.len() > 1 {
```
Comment (nshaaban-cPacket) [inline socketio/src/client/raw_client.rs:398]: nit: cilppy lint
```suggestion
            let payload = if ! contents.is_empty() {
```

File: socketio/src/client/raw_client.rs [hunk 19]
```diff
@@ -306,51 +319,95 @@ impl RawClient {
...
+                Payload::Multi(contents[1..].iter().map(|v| v.to_owned().into()).collect())
```
Comment (nshaaban-cPacket) [inline socketio/src/client/raw_client.rs:399]: nit: alternative. No difference at all, just if you feel like using iterator's functions
```suggestion
                Payload::Multi(contents.iter().skip(1).cloned().map(RawPayload::from).collect())
```

File: socketio/src/client/raw_client.rs [hunk 20]
```diff
@@ -306,51 +319,95 @@ impl RawClient {
...
+        // the string must be a valid json array with the event at index 0 and the
+        // payload at index 1. if no event is specified, the message callback is used
+        if let Some(serde_json::Value::Array(contents)) = &packet.data {
+            let event: Event = if contents.len() > 1 {
+                contents
+                    .get(0)
+                    .map(|value| match value {
+                        serde_json::Value::String(ev) => ev,
+                        _ => "message",
+                    })
```
Comment (nshaaban-cPacket) [inline socketio/src/client/raw_client.rs:391]:
```suggestion
                    .and_then(|value| match value {
                        serde_json::Value::String(ev) => Some(ev),
                        _ => None
                    })
```

**Review by nshaaban-cPacket — COMMENTED** [review]
Thank you for being receptive to comments so far, feel free to ignore the "nit"s you don't want to do, just saying things I noticed.

Changes to Packet's public interface require @1c3t3a 's approval. We MIGHT be able to still squeeze by with changes to Packet, as it's not likely many use it. Your call.

File: socketio/src/payload.rs (hunk 9 above)
Comment (SSebo) [reply socketio/src/payload.rs:12]: I tried `Multi(Vec<Payload>)` first time, recursive actually can be convert to valid json, but not what we expected. To not let user do that, `RawPayload` is provided. also let code simple.

File: socketio/src/socket.rs (hunk 10 above)
Comment (SSebo) [reply socketio/src/socket.rs:101]: I found  `client.emit(Buffer.from([4, 5, 6]));` in socket-io.js, I thought event is optional that time. But I lookup the socketio dos, event is not optional. Does this mean the event can be a valid payload?

File: socketio/src/packet.rs (hunk 13 above)
Comment (SSebo) [reply socketio/src/packet.rs:228]: Yes, _placeholder will be replaced to real `Payload::Binary` in socket.rs

File: socketio/src/packet.rs (hunk 14 above)
Comment (SSebo) [reply socketio/src/packet.rs:24]: Since the `Payload::Multi` is a breaking change, so I change data to `Value`.

File: socketio/src/client/raw_client.rs (hunk 18 above)
Comment (SSebo) [reply socketio/src/client/raw_client.rs:398]: `len() > 0`  is equal to `! is_empty`. Here is `len() > 1`, is just one item, no need to use `Payload::Multi`

File: socketio/src/socket.rs (hunk 10 above)
Comment (SSebo) [reply socketio/src/socket.rs:101]: right now in `rust-socketio` if only receive 1 item in packet.data, it will be treat as payload, not event.

https://github.com/1c3t3a/rust-socketio/blob/4ac27c7158a89375c503164857d86b1a1167116c/socketio/src/client/raw_client.rs#L325-L354

File: socketio/src/client/raw_client.rs (hunk 17 above)
Comment (SSebo) [reply socketio/src/client/raw_client.rs]: here is len() == 1

File: socketio/src/socket.rs (hunk 10 above)
Comment (1c3t3a) [reply socketio/src/socket.rs:101]: Where did you find that? Specification only mentions, that when no event is specified, "Message" should be used. So having an `Option<Event>` would not make sense in that case.

File: socketio/src/payload.rs (hunk 9 above)
Comment (1c3t3a) [reply socketio/src/payload.rs:12]: I think this makes it a bit obvious that restructuring payload is probably a bad idea. What you do here is basically rewrite a limited Json `Value` enum, with support for binary data. I would rather leverage json, and solve this problem differently. I am going to take a look at this tomorrow, but I am not really in favor of such changes, as they present obvious problems.

File: socketio/src/packet.rs (hunk 13 above)
Comment (nshaaban-cPacket) [reply socketio/src/packet.rs:227]: packet.rs should handle all de-serialization of the packet. This will become more important if/when we want to support multiple versions.

File: socketio/src/client/raw_client.rs (hunk 17 above)
Comment (nshaaban-cPacket) [reply socketio/src/client/raw_client.rs]: Yes hence why it's `! vec.is_empty()`. is_empty() is as faster operation than len()

File: socketio/src/socket.rs [hunk 21]
```diff
@@ -158,6 +158,47 @@ impl Socket {
         attachments.push(bin_data);
     }
 
+    pub(crate) fn decode_binary_payload(
```
Comment (nshaaban-cPacket) [inline socketio/src/socket.rs:161]: nit: why did this move? It's only used in raw_client

Comment (SSebo) [reply socketio/src/socket.rs:161]: I found encode is in socket.rs, so I move decode in socket.rs too.

File: socketio/src/socket.rs (hunk 10 above)
Comment (SSebo) [reply socketio/src/socket.rs:101]: @1c3t3a I found it here https://github.com/1c3t3a/rust-socketio/blob/4ac27c7158a89375c503164857d86b1a1167116c/ci/socket-io.js#L41

File: socketio/src/client/raw_client.rs (hunk 17 above)
Comment (SSebo) [reply socketio/src/client/raw_client.rs]: I got that, if not empty, get(0) will not panic.

File: socketio/src/packet.rs (hunk 13 above)
Comment (SSebo) [reply socketio/src/packet.rs:227]: `BinaryEvent` can be mixed with normal data and binary data like `client.emit('test', Buffer.from([1, 2, 3]), "4", Buffer.from([5, 6]));`, so we can not just remove all `_placeholder` as we do not know where the binary data is in the sequence of received items. `socket.rs` not parse the packet, just replace `_placehoulder` to `Payload::Binary`

**Comment by nshaaban-cPacket** [comment]
Why was this closed @SSebo ?

**Comment by 1c3t3a** [comment]
Good question!
